File: server/app/core/envelope.py
"""Secret envelope format shared with piso/internal/technocore (Go).

Payload: ChaCha20-Poly1305 (RFC 8439) with AAD piso-secret-v1.
Wrap: ephemeral X25519 + HKDF-SHA256 + ChaCha20-Poly1305 of the DEK.

Wrap blob:
  version (1) || eph_pub (32) || nonce (12) || sealed_dek (48)
"""

from __future__ import annotations

import os

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

ENVELOPE_VERSION = 1
ENVELOPE_ALG = "chacha20poly1305-v1"
WRAP_INFO = b"piso-envelope-wrap-v1"
PAYLOAD_AAD = b"piso-secret-v1"
X25519_SIZE = 32
NONCE_SIZE = 12
DEK_SIZE = 32
TAG_SIZE = 16
WRAP_LEN = 1 + X25519_SIZE + NONCE_SIZE + DEK_SIZE + TAG_SIZE


def seal_payload(plaintext: bytes) -> tuple[bytes, bytes, bytes]:
    dek = os.urandom(DEK_SIZE)
    nonce = os.urandom(NONCE_SIZE)
    ct = ChaCha20Poly1305(dek).encrypt(nonce, plaintext, PAYLOAD_AAD)
    return dek, nonce, ct


def open_payload(dek: bytes, nonce: bytes, ciphertext: bytes) -> bytes:
    return ChaCha20Poly1305(dek).decrypt(nonce, ciphertext, PAYLOAD_AAD)


def wrap_dek(recipient_pub: bytes, dek: bytes) -> bytes:
    if len(dek) != DEK_SIZE:
        raise ValueError("dek must be 32 bytes")
    if len(recipient_pub) != X25519_SIZE:
        raise ValueError("recipient public key must be 32 bytes")
    eph = X25519PrivateKey.generate()
    shared = eph.exchange(X25519PublicKey.from_public_bytes(recipient_pub))
    wrap_key = HKDF(
        algorithm=hashes.SHA256(),
        length=DEK_SIZE,
        salt=None,
        info=WRAP_INFO,
    ).derive(shared)
    nonce = os.urandom(NONCE_SIZE)
    sealed = ChaCha20Poly1305(wrap_key).encrypt(nonce, dek, WRAP_INFO)
    eph_pub = eph.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    return bytes([ENVELOPE_VERSION]) + eph_pub + nonce + sealed


def unwrap_dek(recipient_priv: bytes, blob: bytes) -> bytes:
    if len(blob) != WRAP_LEN:
        raise ValueError(f"wrap blob length {len(blob)}, want {WRAP_LEN}")
    if blob[0] != ENVELOPE_VERSION:
        raise ValueError(f"unsupported wrap version {blob[0]}")
    eph_pub = blob[1 : 1 + X25519_SIZE]
    nonce = blob[1 + X25519_SIZE : 1 + X25519_SIZE + NONCE_SIZE]
    sealed = blob[1 + X25519_SIZE + NONCE_SIZE :]
    priv = X25519PrivateKey.from_private_bytes(recipient_priv)
    shared = priv.exchange(X25519PublicKey.from_public_bytes(eph_pub))
    wrap_key = HKDF(
        algorithm=hashes.SHA256(),
        length=DEK_SIZE,
        salt=None,
        info=WRAP_INFO,
    ).derive(shared)
    return ChaCha20Poly1305(wrap_key).decrypt(nonce, sealed, WRAP_INFO)


def open_envelope(recipient_priv: bytes, wrap_blob: bytes, nonce: bytes, ciphertext: bytes) -> bytes:
    dek = unwrap_dek(recipient_priv, wrap_blob)
    return open_payload(dek, nonce, ciphertext)

File: internal/technocore/envelope.go
package technocore

import (
	"crypto/ecdh"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
)

const (
	EnvelopeVersion byte = 0x01
	EnvelopeAlg          = "chacha20poly1305-v1"
	wrapInfo             = "piso-envelope-wrap-v1"
	payloadAAD           = "piso-secret-v1"
	x25519Size           = 32
)

// Wrap blob: version (1) || ephemeral_x25519_pub (32) || nonce (12) || sealed_dek (48)

// SealPayload encrypts plaintext with a random 32-byte DEK (ChaCha20-Poly1305).
func SealPayload(plaintext []byte) (dek, nonce, ciphertext []byte, err error) {
	dek = make([]byte, chacha20poly1305.KeySize)
	if _, err = io.ReadFull(rand.Reader, dek); err != nil {
		return nil, nil, nil, err
	}
	aead, err := chacha20poly1305.New(dek)
	if err != nil {
		return nil, nil, nil, err
	}
	nonce = make([]byte, chacha20poly1305.NonceSize)
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, nil, nil, err
	}
	ciphertext = aead.Seal(nil, nonce, plaintext, []byte(payloadAAD))
	return dek, nonce, ciphertext, nil
}

// OpenPayload decrypts a payload sealed by SealPayload.
func OpenPayload(dek, nonce, ciphertext []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(dek)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, ciphertext, []byte(payloadAAD))
}

// WrapDEK encrypts dek to recipient's X25519 public key.
func WrapDEK(recipientPubHex string, dek []byte) ([]byte, error) {
	if len(dek) != chacha20poly1305.KeySize {
		return nil, fmt.Errorf("dek must be %d bytes", chacha20poly1305.KeySize)
	}
	pubRaw, err := hex.DecodeString(recipientPubHex)
	if err != nil || len(pubRaw) != x25519Size {
		return nil, fmt.Errorf("invalid recipient encryption public key")
	}
	recipient, err := ecdh.X25519().NewPublicKey(pubRaw)
	if err != nil {
		return nil, fmt.Errorf("invalid recipient encryption public key: %w", err)
	}
	eph, err := ecdh.X25519().GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	shared, err := eph.ECDH(recipient)
	if err != nil {
		return nil, err
	}
	wrapKey, err := deriveWrapKey(shared)
	if err != nil {
		return nil, err
	}
	aead, err := chacha20poly1305.New(wrapKey)
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, chacha20poly1305.NonceSize)
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}
	sealed := aead.Seal(nil, nonce, dek, []byte(wrapInfo))
	out := make([]byte, 0, 1+x25519Size+len(nonce)+len(sealed))
	out = append(out, EnvelopeVersion)
	out = append(out, eph.PublicKey().Bytes()...)
	out = append(out, nonce...)
	out = append(out, sealed...)
	return out, nil
}

// UnwrapDEK opens a blob produced by WrapDEK.
func UnwrapDEK(recipientPriv *ecdh.PrivateKey, blob []byte) ([]byte, error) {
	want := 1 + x25519Size + chacha20poly1305.NonceSize + chacha20poly1305.KeySize + chacha20poly1305.Overhead
	if len(blob) != want {
		return nil, fmt.Errorf("wrap blob length %d, want %d", len(blob), want)
	}
	if blob[0] != EnvelopeVersion {
		return nil, fmt.Errorf("unsupported wrap version %d", blob[0])
	}
	off := 1
	ephPub, err := ecdh.X25519().NewPublicKey(blob[off : off+x25519Size])
	if err != nil {
		return nil, fmt.Errorf("ephemeral public key: %w", err)
	}
	off += x25519Size
	nonce := blob[off : off+chacha20poly1305.NonceSize]
	off += chacha20poly1305.NonceSize
	sealed := blob[off:]
	shared, err := recipientPriv.ECDH(ephPub)
	if err != nil {
		return nil, err
	}
	wrapKey, err := deriveWrapKey(shared)
	if err != nil {
		return nil, err
	}
	aead, err := chacha20poly1305.New(wrapKey)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, sealed, []byte(wrapInfo))
}

func deriveWrapKey(shared []byte) ([]byte, error) {
	return hkdf.Key(sha256.New, shared, nil, wrapInfo, chacha20poly1305.KeySize)
}

// OpenEnvelope is the gateway-side helper: unwrap DEK then decrypt payload.
func OpenEnvelope(id Identity, wrapBlob, nonce, ciphertext []byte) ([]byte, error) {
	priv, err := id.EncryptionPrivate()
	if err != nil {
		return nil, err
	}
	dek, err := UnwrapDEK(priv, wrapBlob)
	if err != nil {
		return nil, err
	}
	return OpenPayload(dek, nonce, ciphertext)
}
